src/winrt/activation/catalog.cpp

```diff
@@ -70,26 +76,57 @@ static unordered_map < wstring, shared_ptr<component> > g_types;
 
 HRESULT WinRTLoadComponent(PCWSTR manifest_path)
 {
+    ComPtr<IStream> fileStream;
+    ComPtr<IXmlReader> xmlReader;
+    XmlNodeType nodeType;
+    const WCHAR* localName = L"";
     auto this_component = make_shared<component>();
-    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
-
-    // first line is the name of the DLL that contains the types
-    ifstream manifest_file(manifest_path, ios_base::in);
-    std::string u8_module_name;
-    getline(manifest_file, u8_module_name);
-    this_component->module_path = converter.from_bytes(u8_module_name);
-
-    // second line is the name of the winmd that contains the types (in case of CLR, can be the same as dll)
-    string u8_metadata_path;
-    getline(manifest_file, u8_metadata_path);
-    this_component->metadata_path = converter.from_bytes(u8_metadata_path);
-
-    // each line after is an activatable class. No apartment-specific registration. Implicitly "both".
-    string u8_class_name;
-    while (getline(manifest_file, u8_class_name))
+
+    RETURN_IF_FAILED(SHCreateStreamOnFileEx(manifest_path, STGM_READ, FILE_ATTRIBUTE_NORMAL, false, nullptr, &fileStream));
+    RETURN_IF_FAILED(CreateXmlReader(__uuidof(IXmlReader), (void**)&xmlReader, nullptr));
+    RETURN_IF_FAILED(xmlReader->SetInput(fileStream.Get()));
+    const WCHAR* fileName;
+    while (S_OK == xmlReader->Read(&nodeType))
     {
-        std::wstring wide_class_name = converter.from_bytes(u8_class_name);
-        g_types[wide_class_name] = this_component;
+        if (nodeType == XmlNodeType_Element)
+        {
+            RETURN_IF_FAILED((xmlReader->GetLocalName(&localName, nullptr)));
+            if (localName && _wcsicmp(localName, L"file") == 0)
```

Comment:
The pointer check isn't needed as GetLocalName will return a valid string in all cases (even if it's an empty string).